optics-core/src/Optics/Label.hs

```diff
@@ -87,9 +106,196 @@
+-- more) features with labels as optics, just with a slightly more verbose
+-- syntax.
+--
+-- == The solution
+--
+-- === Prefixless fields with DuplicateRecordFields
+--
+-- If we look at section showing sample usage, you'll see that field names were
+-- prefixed with full name of a data type, whereas defined labels were
+-- prefixless. This is the simplest approach of using labels as optics that does
+-- not require usage of DuplicateRecordFields, but not the optimal one. We
+-- necessarily want field names to be prefixless, i.e. @field@ to be a field
+-- name and @#field@ to be an overloaded label that becomes an optic refering to
+-- this field in the appropriate context.
+--
+-- With this approach we get a working
+--
+-- - autocompletion and
+--
+-- - jump-to-definition
+--
+-- in editors supporting ctags/etags in combination with
+-- <https://hackage.haskell.org/package/hasktags hasktags>, both of which
+-- (especially the latter) are very important for developer's productivity in
+-- real-world code bases.
+--
+-- To illustrate let's look at data types defined with this approach in mind:
+--
+-- @
+-- {-\# LANGUAGE DuplicateRecordFields \#-}
+--
+-- import Data.Time
+--
+-- data User = User { id     :: Int
+--                  , name   :: String
+--                  , joined :: UTCTime
+--                  , movies :: [Movie]
+--                  }
+--
+-- data Movie = Movie { id          :: Int
+--                    , name        :: String
+--                    , releaseDate :: UTCTime
+--                    }
+-- @
+--
+-- Then appropriate 'LabelOptic' instances can be either written by hand or
+-- generated using Template Haskell functions (defined in @Optics.TH@ module
+-- from @optics-th@ package) with
+--
+-- @
+-- makeFieldLabelsWith noPrefixFieldLabels ''User
+-- makeFieldLabelsWith noPrefixFieldLabels ''Movie
+-- @
+--
+-- /Note:/ there exists a similar approach that involves prefixing field names
+-- with the underscore and generation of lenses as ordinary functions so that
+-- @_field@ is the ordinary field name and @field@ is the lens referencing
+-- it. The drawback of such solution is inability to get working
+-- jump-to-definition for field names, which makes navigation in unfamiliar code
+-- bases significantly harder, so it's not recommended.
+--
+-- === Emulation of NoFieldSelectors
 --
+-- Prefixless fields (especially ones with common names such as @id@ or @name@)
+-- leak into global namespace as accessor functions and can generate a lot of
+-- name clashes. Before NoFieldSelectors is available, this can be alleviated by
+-- splitting modules defining types into two, namely:
 --
--- == Structure of 'LabelOptic' instances
+-- 1) A private one that exports full type definitions, i.e. with their fields
+-- and constructors.
+--
+-- 2) A public one that exports only constructors (or no constructors at all if
+-- the data type in question is opaque).
```

Comment:
Hmm, this may be less convenient in a possible future where `setField` and `RecordDotSyntax` are available, because the fields not being in scope will prevent `HasField` constraints being solved (for data abstraction reasons). Though I guess we can cross that bridge if we get to it.

Perhaps we should say something about data abstraction... a consequence of `LabelOptic` (and similar class-based approaches) is that it is not possible to hide the presence of a field for which you have defined a `LabelOptic` instance.

Reply:
> Hmm, this may be less convenient in a possible future where setField and RecordDotSyntax are available, because the fields not being in scope will prevent HasField constraints being solved (for data abstraction reasons).

Right, but with RecordDotSyntax (well, NoFieldSelectors really) you won't need this trick, it's just a design pattern (i.e. workaround ;) until these are available.

> Though I guess we can cross that bridge if we get to it.

Agreed.

> Perhaps we should say something about data abstraction... a consequence of LabelOptic (and similar class-based approaches) is that it is not possible to hide the presence of a field for which you have defined a LabelOptic instance.

I added a paragraph that explain the issue below.